lib/puppet/ssl/host.rb

```diff
@@ -419,6 +430,141 @@ def load_crls(path)
     end
   end
 
+  # Ensures that the CA certificate is available for either generating or
+  # validating the host's cert.
+  # It will first check if the cert is present in memory (used for testing),
+  # then check on disk, and finally try to download it.
+  # @return true if the CA certificate was found, false otherwise
+  def ensure_ca_certificate
+    file_path = certificate_location(CA_NAME)
+    if check_for_certificate_in_memory(CA_NAME)
+      true
+    elsif Puppet::FileSystem.exist?(file_path)
+      load_certificate_bundle(Puppet::FileSystem.read(file_path))
+      true
+    else
+      bundle = download_ca_certificate_bundle
+      if bundle
+        save_certificate_bundle(bundle)
+        true
+      else
+        false
+      end
+    end
+  end
+
+  # Creates an arry of SSL Certificate objects from a PEM-encoding string
+  # of one or more certs.
+  # @param [String] bundle_string PEM-encoded string of certs
+  # @return [[OpenSSL::X509::Certificate]] the certs loaded from the input string
+  def load_certificate_bundle(bundle_string)
+    delimiters = /-----BEGIN CERTIFICATE-----.*?-----END CERTIFICATE-----/m
+    bundle_string.scan(delimiters).map do |cert|
+      OpenSSL::X509::Certificate.new(cert)
+    end
+  end
+
+  # Fetches the CA certificate bundle from the CA server
+  # @return [[OpenSSL::X509::Certificate]] the certs loaded from the response
+  def download_ca_certificate_bundle
+    return nil if Puppet::SSL::Host.ca_location != :remote
+
+    response = Puppet::Rest::Routes.get_certificate(http_client, CA_NAME)
+    if response.ok?
+      body = response.read_body
+      load_certificate_bundle(body)
+    else
+      nil
+    end
+  end
+
+  # Saves the given certs to disc, to a location determined based
+  # on this host's configuration.
+  # @param [[OpenSSL::X509::Certificate]] the certs to save
+  def save_certificate_bundle(cert_bundle)
+    bundle_string = ''
+    cert_bundle.each do |cert|
+      bundle_string += cert.to_s
```

Comment:
I think `to_s` is an alias of `to_pem`, I don't know if there's value in one over the other, but it might be nice to call something that other objects in the system *can't* by default masquerade as.

Also: What's the most idiomatic way to do this? I don't think that modifying a local string in an each block is. I would do `cert_bundle.map(&:to_pem).join("\n")` but that might just be a more clojure-y way of thinking about it.

Reply:
Not sure about what would be most idiomatic. I'm aware of the `&:method` thing, but I haven't really seen it anywhere. That being said, not opposed to that syntax.